Restore Options changes (revert #1722) (#1723)
Qiskit/documentation: docs/guides/runtime-options-overview.mdx

For V1 primitives, if you do not specify a value for an option, the Qiskit Runtime client's default value is used. In V1, the default value is dependant on the code version.

The tables in the [Options classes summary](#options-classes) section lists the default values.

<span id="options-precedence"></span>
### Precedence rules

There are several ways to set options.  If you set different values for the same option in multiple places, this is the order of precedence:

1. Any options set after primitive initialization.
2. Any options set during primitive initialization.

<Admonition type="caution" title="Important">
Shots (Sampler) or precision (Estimator) set in `run()` are not considered options, but take precidence over values set for `default_shots`or `default_precision`. See the [Run() method](#run-method) section for full details.
</Admonition>

In the following example, these values would be used for the `estimator.run` job: `shots = 128` and `optimization_level = 1`.

``` python
# Setting options during primitive initialization

> estimator = Estimator(backend, options={"optimization_level": 0})

Reviewer: let's not use `optimization_level` since it's deprecated.

# Setting options after primitive initialization
# This uses auto complete.
estimator.options.optimization_level = 1
estimator.options.default_shots = 4000

# Sample two circuits at 128 shots each.

> estimator.run([circuit1, circuit2], shots=128)

Reviewer: this syntax is incorrect. You need observables in estimator.

```suggestion
estimator.run([(circuit1, observable1), (circuit2, observable2)], shots=128)
```

```

#### Special cases

The `default_shots` and `resilience_level` options require special consideration:

If you set `default_shots` in an estimator, it overrides any value set for `default_precision`.

The `resilience_level` is used to set a base configuration on any relevant options, and then any other user-specified options override that configuration. So it would be possible to undo the effect of setting the resilience level by reversing all of its settings manually.

In the following example, setting the resilience level to 0 does not turn off `zne_mitigation` because `estimator.options.resilience.zne_mitigation = True` overrides the relevent setup from `estimator.options.resilience_level = 0`.

``` python
from qiskit_ibm_runtime import EstimatorV2, QiskitRuntimeService
from qiskit import QuantumCircuit

service = QiskitRuntimeService()
backend = service.backend("ibm_auckland")

estimator = EstimatorV2(backend)

estimator.options.default_shots = 100
estimator.options.resilience_level = 0
estimator.options.resilience.zne_mitigation = True
```

<span id="options-classes"></span>
## Options classes summary

<Tabs>
  <TabItem value="EstimatorV2" label="Estimator V2">

>   - [Resilience](../api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.ResilienceOptionsV2): Advanced options for configuring error mitigation methods such as measurement error mitigation, ZNE, and PEC. Estimator only.

Reviewer: It doesn't really need `Estimator only` since it's already listed under EstimatorV2 tab

```
</Admonition>

## Overview

<span id="options-structure"></span>
### Structure

When calling the primitives, you can pass in options by using an options class or a dictionary. Commonly used options, such as `resilience_level`, are at the first level. Other options are grouped into different categories, such as `execution`. See the [Set primitive options](specify-runtime-options) topic for full details.

<span id="options-defaults"></span>
### Defaults

For V2 primitives, if you do not specify a value for an option, it is given a special value of `Unset` and the server default value is used. Thus, the default value will be the same regardless of your code version.

> For V1 primitives, if you do not specify a value for an option, the Qiskit Runtime client's default value is used. In V1, the default value is dependant on the code version.

Reviewer 2: ```suggestion
For V1 primitives, if you do not specify a value for an option, the Qiskit Runtime client's default value is used. In V1, the default value is dependent on the code version.
```

<span id="options-classes"></span>
## Options classes summary

<Tabs>
  <TabItem value="EstimatorV2" label="Estimator V2">
  - [Resilience](../api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.ResilienceOptionsV2): Advanced options for configuring error mitigation methods such as measurement error mitigation, ZNE, and PEC.
  - [Dynamical decoupling](../api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.DynamicalDecouplingOptions): Options for dynamical decoupling.
  - [Execution](../api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.ExecutionOptionsV2): Primitive execution options, including whether to initialize qubits and the repetition delay.
  - [Twirling](../api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.TwirlingOptions): Twirling options, such as whether to apply two-qubit gate twirling and the number of shots to run for each random sample.

>   - [Environment](../api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.EnvironmentOptions): Execution environment options such as the logging level to set and job tags to add.

Reviewer 2: ```suggestion
  - [Environment](../api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.EnvironmentOptions): Execution environment options, such as the logging level to set and job tags to add.
```

  - [Simulator](../api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.SimulatorOptions): Simulator options, such as the basis gates, simulator seed, and coupling map. Applies to local testing mode only.
  </TabItem>

  <TabItem value="SamplerV2" label="Sampler V2">
  - [Dynamical decoupling](../api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.DynamicalDecouplingOptions): Options for dynamical decoupling.
  - [Execution](../api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.ExecutionOptionsV2): Primitive execution options, including whether to initialize qubits and the repetition delay.
  - [Twirling](../api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.TwirlingOptions): Twirling options, such as whether to apply two-qubit gate twirling and the number of shots to run for each random sample.

>   - [Environment](../api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.EnvironmentOptions): Execution environment options such as the logging level to set and job tags to add.

Reviewer 2: ```suggestion
  - [Environment](../api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.EnvironmentOptions): Execution environment options, such as the logging level to set and job tags to add.
```

  - [Simulator](../api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.SimulatorOptions): Simulator options, such as the basis gates, simulator seed, and coupling map. Applies to local testing mode only.
  </TabItem>

  <TabItem value="OptionsV1" label="V1 options">
  - [Resilience](../api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.ResilienceOptions): Advanced options for configuring error mitigation methods such as measurement error mitigation, ZNE, and PEC. **Estimator only**.
  - [Dynamical decoupling](../api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.DynamicalDecouplingOptions): Options for dynamical decoupling.
  - [Execution](../api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.ExecutionOptions): Primitive execution options, including whether to initialize qubits and the repetition delay.
  - [Twirling](../api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.TwirlingOptions): Twirling options, such as whether to apply two-qubit gate twirling and the number of shots to run for each random sample.

>   - [Environment](../api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.EnvironmentOptions): Execution environment options such as the logging level to set and job tags to add.

Reviewer 2: ```suggestion
  - [Environment](../api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.EnvironmentOptions): Execution environment options, such as the logging level to set and job tags to add.
```

<span id="options-structure"></span>
### Structure

When calling the primitives, you can pass in options by using an options class or a dictionary. In the options classes, commonly used options, such as `resilience_level`, are at the first level. Other options are grouped into different categories, such as `execution`. See the [Set primitive options](specify-runtime-options) topic for full details.

<span id="options-defaults"></span>
### Defaults

For V2 primitives, if you do not specify a value for an option, it is given a special value of `Unset` and the server default value is used. Thus, the default value will be the same regardless of your code version.
For V1 primitives, if you do not specify a value for an option, the Qiskit Runtime client's default value is used. In V1, the default value is dependant on the code version.

The tables in the [Options classes summary](#options-classes) section lists the default values.

<span id="options-precedence"></span>
### Precedence rules

There are several ways to set options.  If you set different values for the same option in multiple places, this is the order of precedence:

1. Any options set after primitive initialization.

> 2. Any options set during primitive initialization.

Reviewer 3: > . I might rephrase the section title

Do you have a suggestion for a different title, or is there more content we could add here to make it a better section about precedence rules?  I used the title "precedence rules" to address a specific request that we document precedence rules.

> 2. Any options set during primitive initialization.

Reviewer 4: I think talking about "precedence rules" in the context of shots/precision makes sense (but is not strictly necessary) because there there are some (pub > run > default_shots > default_precision), so I'm wondering if there might have been a miscommunication on this front.

Below is a rough outline of how I might organize this section. Please edit however you like, adding snippets if you think necessary, but sometimes less is more, imo.
[and please check my code snippets run because I didn't :)]

## Suggestion

<span id="setting-options"></span>
### Setting Options

Options can be defined before a primitive is constructed and passed to the primitive, which will _make a copy_ of them. As shown below, this can be done either as a nested dictionary, or using the options classes themselves. However, this is not necessary, because once the primitive is constructed, the options that it owns can be mutated. You can choose the workflow that works best for your application.

``` python
# Setting options during primitive initialization with a nested dictionary
estimator1 = Estimator(backend, options={"twirling": {"enable_gates": True}})

# Updating the options after intialization
estimator1.options.twirling.num_randomizations = 100

# Setting options during primitive initialization with an options class
options = EstimatorOptionsV2()
options.twirling.enable_gates = True
estimator2 = Estimator(backend, options=options)

# Updating the options after intialization
estimator2.options.twirling.num_randomizations = 100

# Keep all default options during primitive initialization
estimator3 = Estimator(backend)

# Updating the options after intialization
estimator3.options.twirling.enable_gates = True
estimator3.options.twirling.num_randomizations = 100
```

<span id="special-cases"></span>
### Special cases

#### Resilience level (estimator only)

The resilience level is not so much an option, but a specification of a base set of curated options to build off of. Roughly, resilience level 0 turns all bells and whistles off, resilience level 1 turns on options for measurement error mitigation, and resilience level 2 turns on options for gate and measurement error mitigation. 

Any options you manually specify in addition to the resilience level are applied _on top of_ the base set of options defined by the resilience level. So in principle, you could set the resilience level to 1, but then turn off measurement mitigation, although this would not be advised.

#### Shots (sampler only)

The SamplerV2.run method accepts two arguments: a list of pubs, each of which can specify a pub-specific value of shots, and a shots keyword argument. These shot values are a part of the sampler execution interface, and are independent of the runtime estimator's options---they necessarily take precedence over any values specified inside of the options in order to comply with the sampler abstraction.

However, if shots is not specified by any pub or in the run keyword argument (equivalently, if they are all None), then the shots value will inferred from the options, most notably `default_shots`. To sumarize, this is the the order of precedence for specifying shots in the sampler, for any particular pub:

1. if the pub specifies shots itself, use that, otherwise
2. if the `run` shots keyword argument is specified, use that, otherwise
3. use the value of `sampler.options.default_shots`.

#### Precision (estimator only)

Precision is analogous to shots, described in the previous section, with the addition of the options containing both `default_shots` and `default_precision`. Specifically, for any particular estimator pub:

1. if the pub specifies precision itself, use that, otherwise
2. if the `run` precision keyword argument is specified, use that, otherwise
3. if the `estimator.options.default_shots` argument is specified, use that to control the amount of data, otherwise
4. use the value of `estimator.options.default_precision`

<Admonition type="caution" title="Important">
Shots (Sampler) or precision (Estimator) set in `run()` are not considered options, but take precidence over values set for `default_shots`or `default_precision`. See the [Run() method](specify-runtime-options#run-method) section for full details.
</Admonition>

> In the following example, these values would be used for the `estimator.run` job: `shots = 128` and `resilience_level = 1`.

Reviewer: The code doesn't actually change the resilience level; it just turns on meas mitigation. 

```suggestion
In the following example, these values would be used for the `estimator.run` job: `shots = 128` and `measure_mitigation = True`.
```

``` python
# Setting options during primitive initialization
estimator = Estimator(backend, options={'resilience_level': 2,'resilience': {'measure_mitigation': False}})

# Setting options after primitive initialization
# This uses auto complete.
estimator.options.resilience.measure_mitigation = True
estimator.options.default_shots = 4000

# Sample two circuits at 128 shots each.
estimator.run([(circuit1, observable1), (circuit2, observable2)], shots=128)
```

#### Special cases

The `default_shots` and `resilience_level` options require special consideration:

If you set `default_shots` in an estimator, it overrides any value set for `default_precision`.

> The `resilience_level` is used to set a base configuration on any relevant options, and then any other user-specified options override that configuration. So it would be possible to undo the effect of setting the resilience level by reversing all of its settings manually.

Reviewer 4: ```suggestion
The `resilience_level` is used to set a base configuration on any relevant options. For example, setting the resilience level to 2 in the estimator tells the runtime estimator to start with a base options configuration that enables TREX, gate twirling, and ZNE with sensible defaults. Any other user-specified options override that configuration. So it would be possible to undo the effect of setting the resilience level by reversing all of its settings manually.
```

|                      |                          | shots_per_randomization |  integer                                                                                | 'auto'                    |                      |
|                      | [pec](/api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.PecOptions)   | max_overhead            |  `None`/ integer >1          | `100`                      |                      |
|                      |                          | noise_gain              |  `auto`/ float in the range `[0, 1]`                                                     | `auto`                    |                      |
|                      | pec_mitigation           |                         | `True`/`False`                                                                        | `False`                      |                      |
|                      | zne_mitigation           |                         | `True`/`False`                                                                         | `False`                     |                      |

> |                      | [zne](/api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.ZneOptions)                      | noise_factors           |   list of floats; each float >= 1     | `(1, 3, 5)`           |                      |

Reviewer: nit... `pec_mitigation` should go before `pec`, like how `zne_mitigation` is before `zne`.

| [simulator](../api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.SimulatorOptions)            |                          |                         |                        |                     |   simulator  |
|                      | noise_model              |                         |   [Qiskit Aer NoiseModel](build-noise-models), or its representation        | `None`                      |              |
|                      | seed_simulator           |                         |   integer                                                                               | `None`                    |              |
|                      | coupling_map             |                         |  List of directed two-qubit interactions                                          | full connectivity                   |              |
|                      | basis_gates              |                         |  List of basis gate names to unroll to          | The set of all basis gates supported by [Qiskit Aer simulator](https://qiskit.github.io/qiskit-aer/stubs/qiskit_aer.AerSimulator.html) |              |

  </TabItem>

</Tabs>

<span id="options-compatibility-table"></span>

> ### Options compatibility

Reviewer 4: Note that PEA has not been publicly released, it's still an experimental option. It is planned to be made public at the end of July.

Reviewer 3: @jyu00 - should we pull all the PEA content until July?  I have an issue for that, so I could move this PEA content in there.

Reviewer 3: Jessie said that it's OK to keep here because it is already publicly documented.

```
</Admonition>

## Overview

<span id="options-structure"></span>
### Structure

When calling the primitives, you can pass in options by using an options class or a dictionary. Commonly used options, such as `resilience_level`, are at the first level. Other options are grouped into different categories, such as `execution`. See the [Set primitive options](specify-runtime-options) topic for full details.

<span id="options-defaults"></span>
### Defaults

For V2 primitives, if you do not specify a value for an option, it is given a special value of `Unset` and the server default value is used. Thus, the default value will be the same regardless of your code version.
For V1 primitives, if you do not specify a value for an option, the Qiskit Runtime client's default value is used. In V1, the default value is dependant on the code version.

The tables in the [Options classes summary](#options-classes) section lists the default values.

<span id="options-precedence"></span>

> ## Set options

Reviewer: This entire section applies only to V2 primitives. We should have one for V1, or at least make it explicit.

Reviewer 3: I think I fixed this.

> Options can be defined before a primitive is constructed and passed to the primitive, which will make a copy of them. As shown below, this can be done either as a nested dictionary, or by using the options classes. However, this is not necessary, because after the primitive is constructed, the options that it owns can be changed. You can choose the workflow that works best for your application.

Reviewer: > However, this is not necessary,

It's unclear to me what is unnecessary? Passing options to a primitive prior to constructing it?

``` python
from qiskit_ibm_runtime.options import EstimatorOptions

# Set options during primitive initialization with a nested dictionary
estimator1 = Estimator(backend, options={"twirling": {"enable_gates": True}})

# Update the options after initialization
estimator1.options.twirling.num_randomizations = 100

# Set options during primitive initialization with an options class
options = EstimatorOptions()
options.twirling.enable_gates = True
estimator2 = Estimator(backend, options=options)

# Update the options after initialization
estimator2.options.twirling.num_randomizations = 100

# Keep all default options during primitive initialization
estimator3 = Estimator(backend)

# Update the options after initialization
estimator3.options.twirling.enable_gates = True
estimator3.options.twirling.num_randomizations = 100
```

### Special cases

#### Resilience level (Estimator only)

> The resilience level is not actually an option, but is a specification of a base set of curated options to build off of. In general, level 0 turns off all error mitigation, level 1 turns on options for measurement error mitigation, and level 2 turns on options for gate and measurement error mitigation.

Reviewer: I get where this is coming from, but I think it'd be confusing to say it's not an option when it's inside `EstimatorOption`

```suggestion
The resilience level is not an option that directly impacts the primitive query. It is instead a specification of a base set of curated options to build off of. In general, level 0 turns off all error mitigation, level 1 turns on options for measurement error mitigation, and level 2 turns on options for gate and measurement error mitigation.
```

Reviewer 4: This is one of the reasons I don't like resilience levels. It really, really isn't like the other options, which is why I don't like to think of it as an option.

Any options you manually specify in addition to the resilience level are applied on top of the base set of options defined by the resilience level. So in principle, you could set the resilience level to 1, but then turn off measurement mitigation, although this is not advised.

### Shots (Sampler only)

> The `SamplerV2.run` method accepts two arguments: a list of PUBs, each of which can specify a PUB-specific value for shots, and a shots keyword argument. These shot values are a part of the Sampler execution interface, and are independent of the Runtime Estimator's options.  They take precedence over any values specified as options in order to comply with the Sampler abstraction.

Reviewer: ```suggestion
The `SamplerV2.run` method accepts two arguments: a list of PUBs, each of which can specify a PUB-specific value for shots, and a shots keyword argument. These shot values are a part of the Sampler execution interface, and are independent of the Runtime Sampler's options.  They take precedence over any values specified as options in order to comply with the Sampler abstraction.
```

However, if `shots` is not specified by any PUB or in the run keyword argument (equivalently, if they are all `None`), then the shots value from the options is used, most notably `default_shots`. To summarize, this is the the order of precedence for specifying shots in the Sampler, for any particular PUB:

1. If the PUB specifies shots, use that value.
2. If the shots keyword argument is specified in `run`, use that value.
3. If `sampler.options.default_shots` is specified, use that value.

For example, if shots are specified in all three places, the one with highest precedence (shots specified in the PUB) is used.

### Precision (Estimator only)

> Precision is analogous to shots, described in the previous section, except that the options contains both `default_shots` and `default_precision`. Specifically, for any particular Estimator PUB:

Reviewer: ```suggestion
Precision is analogous to shots, described in the previous section, except that the Estimator options contains both `default_shots` and `default_precision`. Specifically, for any particular Estimator PUB:
```

options.resilience_level = 1
options.execution.shots = 2048
```
<span id="pass-options"></span>
## Set primitive options

You can set options in an options class, when initializing the primitive, or both.

### Options class

> When creating an instance of the `Estimator` or `Sampler` class, you can pass in the `options` you created in the options class. Those options will then be applied when you use `run()` to perform the calculation.  Specify the options in this format:  `primitive.options.option.sub-option.sub-sub-option = choice`.  For example: `estimator.options.dynamical_decoupling.enable = True`

Reviewer: This section is a bit weird. It talks about creating and passing in `options` class, but the code doesn't actually show that (for both V1 and V2). I think the layout for "Set primitive options" should be

- Primitive initialization
    - you can pass in an instance of options class (`EstimatorOptions` and `SamplerOptions` for V2, `Options` for V1), or a dictionary when initializing Sampler/Estimator
- Update options after initialization
    - Both V1 and V2 support the syntax `<primitive>.options.<option>.<sub-option>.<sub-sub-option> = <choice>`, if you want to use auto complete
    - If you want to do bulk update, V1 supports `<primitive>.set_options()`, and V2 supports `<primitive>.options.update()`
- Run() method

Example:

<Tabs>
  <TabItem value="PrimV2" label="V2 primitives">
    `SamplerV2` and `EstimatorV2` have separate options classes that do not need to be instantiated.  You can see the available options and update option values during or after primitive initialization.  Those options will then be applied when you use `run()` to perform the calculation.  Example:

```python

> estimator = Estimator(backend=backend)

Reviewer: this should use `mode` since it's V2

When creating an instance of the `Estimator` or `Sampler` class, you can pass in the `options` you created in the options class. Those options will then be applied when you use `run()` to perform the calculation.  Example:

```python

> estimator = Estimator(session=backend, options=options)

Reviewer: We do support `backend=backend` in V1

  <TabItem value="PrimV1" label="V1 primitives">
You can pass any options to `run()`. Because most users will only overwrite a few options at the job level, it is not necessary to specify the options category. The code below, for example, specifies `shots=1024` instead of `execution={"shots": 1024}` (which is also valid).
```python

> estimator = Estimator(session=backend, options=options)

Reviewer: Same as above

<Tabs>
  <TabItem value="primitivesV2" label="V2 Primitives">
Shots (or precision) can be specified in multiple places in V2.  They are prioritized as follows:

For any Sampler pub:

> 1. Int-valued shots contained in the pub

Reviewer: ```suggestion
1. Integer-valued shots contained in the pub
```

</Tabs>

> ### Error suppression

Reviewer: Since we already have entire pages on error suppression and mitigation, I don't think we need to repeat them here. We can instead add `max_execution_time` as another commonly used option.
